src/krylov_phiv.jl

```diff
@@ -123,8 +123,10 @@ function expv!(w::AbstractVector{Complex{Tw}}, t::Complex{Tt}, Ks::KrylovSubspac
         expH = exponential!(t * cache,expmethod)
         expHe = @view(expH[:, 1])
     end
-    lmul!(beta, mul!(w, @view(V[:, 1:m]), expHe)) # exp(A) ≈ norm(b) * V * exp(H)e
+    lmul!(beta, mul!(w, @view(V[:, 1:m]), match_array_type(V, expHe))) # exp(A) ≈ norm(b) * V * exp(H)e
 end
+# NOTE: this function is for CuArrays, which fall back to identity, i.e. keep the source a dense CPU array as it is.
+match_array_type(target::AbstractArray, source::AbstractArray) = source
```

Comment:
Should this go to https://github.com/JuliaArrays/ArrayInterface.jl ?

Reply:
Good point, I found the following API that will do this trick
```julia
julia> ArrayInterface.restructure(CUDA.randn(2,2),x )
2×2 CuArray{Int64, 2, CUDA.Mem.DeviceBuffer}:
 1  2
 3  4
```

Reply:
Oops, this interface is not the desired one.

Reply:
I reverted the change, and rename the `match_array_type` to `compatible_multiplicative_operand`.
Bceause it is not for matching certain array types, e.g. for sparse arrays, the operand is ok to be dense.

I do think this function should be in `ArrayInterface`, but that PR might require some work to understand different array libraries, let start by posting an issue there.